internal/pubsub/subscriber.go

```diff
@@ -103,167 +90,250 @@ func (s Subscriber) Listen() error {
-
-		return err
+func (s *Subscriber) processEvent(msg events.Request[events.AuthRelationshipRequest, events.AuthRelationshipResponse]) error {
+	elogger := s.logger.With(
+		"event.message.topic", msg.Topic(),
+		"event.message.action", msg.Message().Action,
+		"event.message.object.id", msg.Message().ObjectID.String(),
+		"event.message.relations", len(msg.Message().Relations),
+	)
+
+	if msg.Error() != nil {
+		elogger.Errorw("message contains error:", "error", msg.Error())
+
+		return msg.Error()
 	}
 
-	ctx := events.TraceContextFromChangeMessage(context.Background(), changeMsg)
+	request := msg.Message()
+
+	ctx := request.GetTraceContext(context.Background())
 
-	ctx, span := tracer.Start(ctx, "pubsub.receive", trace.WithAttributes(attribute.String("pubsub.subject", changeMsg.SubjectID.String())))
+	ctx, span := tracer.Start(ctx, "pubsub.receive", trace.WithAttributes(attribute.String("pubsub.subject", request.ObjectID.String())))
 
 	defer span.End()
 
-	resource, err := s.qe.NewResourceFromID(changeMsg.SubjectID)
-	if err != nil {
-		s.logger.Warnw("invalid subject id", "error", err.Error())
+	elogger.Debugw("received message")
 
-		return nil
+	var err error
+
+	switch request.Action {
+	case events.WriteAuthRelationshipAction:
+		err = s.handleCreateEvent(ctx, msg)
+	case events.DeleteAuthRelationshipAction:
+		err = s.handleDeleteEvent(ctx, msg)
+	default:
+		elogger.Warnw("ignoring msg, not a write or delete action")
 	}
 
-	s.logger.Debugw("received message", "resource_type", resource.Type, "resource_id", resource.ID.String(), "event_type", changeMsg.EventType)
+	if err != nil {
+		return err
+	}
 
-	switch events.ChangeType(changeMsg.EventType) {
-	case events.CreateChangeType:
-		err = s.handleCreateEvent(ctx, msg, changeMsg)
-	case events.UpdateChangeType:
-		err = s.handleUpdateEvent(ctx, msg, changeMsg)
-	case events.DeleteChangeType:
-		err = s.handleDeleteEvent(ctx, msg, changeMsg)
-	default:
-		s.logger.Warnw("ignoring msg, not a create, update or delete event", "event_type", changeMsg.EventType)
+	return nil
+}
+
+func (s *Subscriber) createRelationships(ctx context.Context, relationships []types.Relationship) error {
+	// Attempt to create the relationships in SpiceDB.
+	_, err := s.qe.CreateRelationships(ctx, relationships)
+	if err != nil {
+		return fmt.Errorf("%w: error creating relationships", err)
 	}
 
+	return nil
+}
+
+func (s *Subscriber) deleteRelationships(ctx context.Context, relationships []types.Relationship) error {
+	_, err := s.qe.DeleteRelationships(ctx, relationships...)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (s *Subscriber) createRelationships(ctx context.Context, msg *message.Message, resource types.Resource, additionalSubjectIDs []gidx.PrefixedID) error {
-	var relationships []types.Relationship
+func (s *Subscriber) handleCreateEvent(ctx context.Context, msg events.Request[events.AuthRelationshipRequest, events.AuthRelationshipResponse]) error {
```

Comment:
In general, the use of `respondRequest` in so many places here (with the exception of actually creating the relationships) makes me think we actually should just ack any message that comes in and respond with errors if something fails. That way, the client can decide if it wants to retry or just bail. This is particularly true IMO given that both `handleCreateEvent` and `handleDeleteEvent` always return `nil`.

Reply:
Idea: Only call `Nak` if we can't actually reply to the message.

Reply:
good point. i think naking only on reply failure makes sense.

Reply:
updated to make responseRequest return an error when attempting to reply. if the reply fails we should then nak the message and reprocess it.